engine/src/flutter/lib/web_ui/lib/src/engine/compositing/canvas_provider.dart

```diff
@@ -0,0 +1,123 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' as ui;
+
+/// Manages the lifecycle of raw canvas elements, abstracting away the differences
+/// between onscreen and offscreen canvases.
+///
+/// This class is responsible for:
+/// - Acquiring and releasing canvas elements.
+/// - Resizing canvases.
+/// - Attaching `webglcontextlost` event listeners and notifying the consumer.
+abstract class CanvasProvider<C extends DomEventTarget> {
+  final Map<C, ui.VoidCallback> _contextLostCallbacks = <C, ui.VoidCallback>{};
+  final Map<C, DomEventListener> _eventListeners = <C, DomEventListener>{};
+
+  /// Acquires a canvas element of a given `size`.
+  ///
+  /// The `onContextLost` callback will be invoked when the underlying rendering
+  /// context for this canvas is lost.
+  C acquireCanvas(BitmapSize size, {required ui.VoidCallback onContextLost}) {
+    final C canvas = _createCanvas(size);
+    final DomEventListener eventListener = createDomEventListener((DomEvent event) {
+      final ui.VoidCallback? callback = _contextLostCallbacks[canvas];
+      if (callback != null) {
+        callback();
+      }
```

Comment:
There's no need to have two maps. The `_eventListeners` map should be enough, and in here you can access `onContextLost` from the scope:

```suggestion
      onContextLost();
```